### PR TITLE
QuickStarts templates for App Service Environment v3

### PR DESCRIPTION
**Pull request by samlin-msft**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

*
*
*

### CODE REVIEW

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/virtualnetwork copy.bicep [hunk 1]
```diff
@@ -0,0 +1,66 @@
+@description('Required. The Virtual Network (vNet) Name.')
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/virtualnetwork copy.bicep:1]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/nested/azuredeploy.site.parameters copy.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
+{
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/nested/azuredeploy.site.parameters copy.json:1]: done!

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/nested/azuredeploy.site copy.json [hunk 3]
```diff
@@ -0,0 +1,157 @@
+{
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/nested/azuredeploy.site copy.json:1]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/networksecuritygroup.bicep [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        //destinationApplicationSecurityGroups: ((length(item.properties.destinationApplicationSecurityGroups) == 0) ? json('null') : concat(emptyArray, array(json('{{"id": "${resourceId('Microsoft.Network/applicationSecurityGroups', item.properties.destinationApplicationSecurityGroups[0].name)}","location": "${location}"}}'))))
+        destinationPortRanges: ((length(item.properties.destinationPortRanges) == 0) ? json('null') : item.properties.destinationPortRanges)
+        destinationPortRange: ((item.properties.destinationPortRange == '') ? json('null') : item.properties.destinationPortRange)
+        direction: item.properties.direction
+        priority: int(item.properties.priority)
+        protocol: item.properties.protocol
+        sourceAddressPrefix: ((item.properties.sourceAddressPrefix == '') ? json('null') : item.properties.sourceAddressPrefix)
+        //sourceApplicationSecurityGroups: ((length(item.properties.sourceApplicationSecurityGroups) == 0) ? json('null') : concat(emptyArray, array(json('{{"id": "${resourceId('Microsoft.Network/applicationSecurityGroups', item.properties.sourceApplicationSecurityGroups[0].name)}","location": "${location}"}}'))))
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/networksecuritygroup.bicep:31]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/README.md [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+**Az cli:**
+
+```bash
+az group create --name rg-asev3-templates-demo --location westus2
+```
+
+
+
+## Deployment steps
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/README.md:106]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/scripts/deploy.arm.sh [hunk 6]
```diff
@@ -0,0 +1 @@
+az deployment group create --template-file azuredeploy.json --parameters azuredeploy.parameters.json -g rg-asev3-templates-demo --name arm-asev3-deployment-2021081201
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/scripts/deploy.arm.sh:1]: I removed the scripts folder.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/scripts/deploy.bicep.sh [hunk 7]
```diff
@@ -0,0 +1 @@
+az deployment group create --template-file main.bicep --parameters azuredeploy.parameters.json -g rg-asev3-templates-demo --name bicep-asev3-deployment-2021081201
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/scripts/deploy.bicep.sh:1]: I removed the scripts folder.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/nested/azuredeploy.site.parameters.json [hunk 8]
```diff
@@ -0,0 +1,39 @@
+{
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/nested/azuredeploy.site.parameters.json:1]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/nested/azuredeploy.site.parameters copy.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/nested/azuredeploy.site.parameters copy.json:1]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/azuredeploy.parameters.json [hunk 9]
```diff
@@ -0,0 +1,69 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/azuredeploy.parameters.json:4]: remove all hard-coded parameter and leave the aseNamePrefix with "GEN-UNIQUE" value.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep [hunk 10]
```diff
@@ -0,0 +1,165 @@
+@description('Optional. Add unique suffix string to ASE name.')
+param addUniqueSuffix bool = true
+
+@description('Required. Use existing virtual network and subnet.')
+param useExistingVnetandSubnet bool = false
+
+@description('Required. Resource Group name of virtual network if using existing vnet and subnet.')
+param vNetResourceGroupName string = 'rg-asev3-vnet'
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep:8]: modified.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+      }
+    ]
+    privateEndpointNetworkPolicies: 'Enabled'
+    privateLinkServiceNetworkPolicies: 'Enabled'
+    networkSecurityGroupName: 'nsg-asev3'
+  }
+]
+
+@description('Required. Name of ASEv3.')
+param aseNamePrefix string = 'asev3-ilb'
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep:48]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+@description('Required. Array of Security Rules to deploy to the Network Security Group.')
+param networkSecurityGroupSecurityRules array
+
+@description('Optional. It is only for unique string generation base on timestamp.')
+param timeStamp string = utcNow()
+
+// Variable definitions
+var uniStr = substring('${uniqueString(resourceGroup().id, timeStamp)}', 0, 4)
+var aseName = addUniqueSuffix ? '${aseNamePrefix}-${uniStr}' : aseNamePrefix
+var virtualNetworkId = useExistingVnetandSubnet ? resourceId(vNetResourceGroupName, 'Microsoft.Network/virtualNetworks', virtualNetworkName) : resourceId('Microsoft.Network/virtualNetworks', virtualNetworkName)
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep:79]: There is a scenario 2 that virtual network can be place in another resource group. I need condition to deploy different resource group to support this.

- Scenario 1 : Deploy ASEv3 to a new virtual network in the same resource group
- Scenario 2 : Deploy ASEv3 to an existing virtual network in different resource group

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+var subnetId =  useExistingVnetandSubnet ? resourceId(vNetResourceGroupName, 'Microsoft.Network/virtualNetworks/subnets', virtualNetworkName, subnetName) : resourceId('Microsoft.Network/virtualNetworks/subnets', virtualNetworkName, subnetName)
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep:80]: Same reason as above. 

There is a scenario 2 that virtual network can be place in another resource group. I need condition to deploy different resource group to support this.

- Scenario 1 : Deploy ASEv3 to a new virtual network in the same resource group
- Scenario 2 : Deploy ASEv3 to an existing virtual network in different resource group

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+var privateDNSZoneName = asev3.properties.dnsSuffix
+var aseNetworkConfiguration = '${asev3.id}/configurations/networking'
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep:82]: line 82: asev3.id is the bicep resource id, will be converted to resource id in arm template.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+
+@description('Required. Name of the Network Security Group.')
+@minLength(1)
+param networkSecurityGroupName string = 'nsg-asev3'
+
+@description('Required. Array of Security Rules to deploy to the Network Security Group.')
+param networkSecurityGroupSecurityRules array
+
+@description('Optional. It is only for unique string generation base on timestamp.')
+param timeStamp string = utcNow()
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep:74]: remove this and "addUniqueSuffix" parameter and "aseName" variables refer to this parameter.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+@description('Required. The subnet Name of ASEv3.')
+param subnetName string = 'snet-asev3'
+
+@description('Required. The subnet properties.')
+param subnets array = [
+  {
+    name: 'snet-asev3'
+    addressPrefix: '172.16.0.0/24'
+    // @description('Required. Delegation name of the ASEv3 subnet.')
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep:32]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+    subnets: [for item in subnets: {
+      name: item.name
+      properties: {
+        addressPrefix: item.addressPrefix
+        networkSecurityGroup: (empty(item.networkSecurityGroupName) ? json('null') : json('{"id": "${resourceId('Microsoft.Network/networkSecurityGroups', item.networkSecurityGroupName)}"}'))
+        delegations: item.delegations
+      }
+    }]
+  }
+  dependsOn: [
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep:124]: subnet addressPrefix and delegations properties are required for ASE deployment, if I remove the properties, template validation will be failed.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/networksecuritygroup.bicep [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+@minLength(1)
+param networkSecurityGroupName string
+
+@description('Optional. Location for all resources.')
+param location string = resourceGroup().location
+
+@description('Required. Array of Security Rules to deploy to the Network Security Group.')
+param networkSecurityGroupSecurityRules array
+
+//var emptyArray = []
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/networksecuritygroup.bicep:11]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/networksecuritygroup.bicep [hunk 19]
```diff
@@ -0,0 +1,37 @@
+@description('Required. Name of the Network Security Group.')
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/networksecuritygroup.bicep:1]: resource networkSecurityGroupName_resource 'Microsoft.Network/networkSecurityGroups@2020-11-01' = {
  name: networkSecurityGroupName <== here
  location: location

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/nested/azuredeploy.site.json [hunk 20]
```diff
@@ -0,0 +1,157 @@
+{
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/nested/azuredeploy.site.json:1]: remove the nested folder.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/virtualnetwork.bicep [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+param networkSecurityGroupName string
+
+@description('Required. Array of Security Rules to deploy to the Network Security Group.')
+param networkSecurityGroupSecurityRules array = []
+
+@description('Required. An Array of subnets to deploy to the Virual Network.')
+@minLength(1)
+param subnets array
+
+// @description('Optional. DNS Servers associated to the Virtual Network.')
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/virtualnetwork.bicep:21]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/virtualnetwork.bicep [hunk 22]
```diff
@@ -0,0 +1,66 @@
+@description('Required. The Virtual Network (vNet) Name.')
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/virtualnetwork.bicep:1]: resource virtualnetwork 'Microsoft.Network/virtualNetworks@2020-11-01' = {
  name: virtualNetworkName <== here
  location: location

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep [hunk 23]
```diff
@@ -0,0 +1,84 @@
+@description('Required. Subscription id.')
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:1]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/privatednszone.bicep [hunk 24]
```diff
@@ -0,0 +1,65 @@
+@description('Required. Private DNS zone name.')
+param privateDNSZoneName string
+
+@description('Required. Network Id.')
+param virtualNetworkId string
+
+@description('Required. ASE network configuration.')
+param aseNetworkConfiguration string
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/privatednszone.bicep:8]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+param numberOfWorkers string = '1'
+@description('Required. Current stack of App service.')
+param currentStack string = 'dotnet'
+@description('Required. Enable php of App service.')
+param phpVersion string = 'OFF'
+@description('Required. .NET Framework version of App service.')
+param netFrameworkVersion string = 'v5.0'
+
+@description('Optional. It is only for unique string generation base on timestamp.')
+param timeStamp string = utcNow()
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:35]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+
+// Variable definitions
+var uniStr = substring('${uniqueString(resourceGroup().id, timeStamp)}', 0, 4)
+var appName = '${appNamePrefix}-${uniStr}'
+var hostingPlanName = '${hostingPlanNamePrefix}-${uniStr}'
+var hostingEnvironmentProfile = {
+  id: resourceId('Microsoft.Web/hostingEnvironments', hostingEnvironmentProfileName)
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:42]: It can be another resource group. Add a new "hostingEnvironmentResourceGroup" parameter for other resource group scenario.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep [hunk 27]
```diff
@@ -0,0 +1,165 @@
...
+  // @description('Required. Three key properties for ASEv3.')
+  properties: {
+    dedicatedHostCount: dedicatedHostCount
+    zoneRedundant: zoneRedundant
+    internalLoadBalancingMode: internalLoadBalancingMode
+    virtualNetwork: {
+      id: subnetId
+    }
+  }
+  dependsOn: [
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep:152]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/networksecuritygroup.bicep [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+  location: location
+  properties: {
+    securityRules: [for item in networkSecurityGroupSecurityRules: {
+      name: item.name
+      properties: {
+        description: item.properties.description
+        access: item.properties.access
+        destinationAddressPrefix: ((item.properties.destinationAddressPrefix == '') ? json('null') : item.properties.destinationAddressPrefix)
+        destinationAddressPrefixes: ((length(item.properties.destinationAddressPrefixes) == 0) ? json('null') : item.properties.destinationAddressPrefixes)
+        //destinationApplicationSecurityGroups: ((length(item.properties.destinationApplicationSecurityGroups) == 0) ? json('null') : concat(emptyArray, array(json('{{"id": "${resourceId('Microsoft.Network/applicationSecurityGroups', item.properties.destinationApplicationSecurityGroups[0].name)}","location": "${location}"}}'))))
```
[the remark this answers is not in the document]
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/networksecuritygroup.bicep:24]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep (hunk 12 above)
[the remark this answers is not in the document]
Comment (bmoore-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep:69]: ```resourceId(vNetResourceGroupName, 'Microsoft.Network/virtualNetworks', virtualNetworkName) ```

will work in all cases since the defaultValue is the current resourceGroup - same for the subnetId below

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep [hunk 29]
```diff
@@ -121,15 +110,12 @@ resource virtualnetwork 'Microsoft.Network/virtualNetworks@2020-11-01' = if (!us
       }
     }]
   }
-  dependsOn: [
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep]: revert this change and move the ```dependsOn``` property above the ```properties``` property (i.e. put it back and change the order)

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep (hunk 27 above)
[the remark this answers is not in the document]
Comment (bmoore-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep]: you need to change the order, not remove.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/virtualnetwork.bicep (hunk 22 above)
[the remark this answers is not in the document]
Comment (bmoore-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/virtualnetwork.bicep:1]: the file itself - doesn't seem to be used anywher

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/privatednszone.bicep [hunk 30]
```diff
@@ -5,7 +5,9 @@ param privateDNSZoneName string
 param virtualNetworkId string
 
 @description('Required. ASE network configuration.')
-param aseNetworkConfiguration string
+param aseNetworkConfigurationId string
+
+var aseNetworkConfigurationEndpoint = '${aseNetworkConfigurationId}/configurations/networking'
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/privatednszone.bicep:10]: use the existing syntax for this, e.g.
```bicep
resource aseConfig 'Microsoft.Web/hostingEnvironments/configurations@2021-01-15' existing = {
name: '${aseName}/networking'
}
```

And then replace the references below with:

```aseConfig.internalInboundIpAddresses[0]```

You'll need to pass in the name of the hosting environment instead of the resourceID

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep [hunk 31]
```diff
@@ -0,0 +1,87 @@
...
+
+var uniStr = '${uniqueString(resourceGroup().id)}'
+var appName = '${appNamePrefix}-${uniStr}'
+var hostingPlanName = '${hostingPlanNamePrefix}-${uniStr}'
+
+resource site 'Microsoft.Web/sites@2021-01-15' = {
+  name: appName
+  location: location
+  properties: {
+    name: appName
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:48]: remove extraneous property

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep [hunk 32]
```diff
@@ -0,0 +1,87 @@
+@description('Required. App service prefix.')
+param appNamePrefix string = 'app-2'
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:2]: remove defaultValue

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep [hunk 33]
```diff
@@ -0,0 +1,87 @@
...
+
+@description('Required. App service location.')
+param location string = resourceGroup().location
+
+@description('Required. App service plan prefix.')
+param hostingPlanNamePrefix string = 'asev3-asp-2'
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:8]: remove defaultValue

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep [hunk 34]
```diff
@@ -0,0 +1,87 @@
...
+
+@description('Required. App service plan hosting environment profile name (ASEv3 name).')
+param hostingEnvironmentProfileName string = 'asev3-ilb-cua3tpryw2j64'
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:11]: remove defaultValue

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep [hunk 35]
```diff
@@ -0,0 +1,87 @@
...
+
+@description('Required. Enable Always-on of App service.')
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:13]: remove "required" from the description for all parameters (all params are required)

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep [hunk 36]
```diff
@@ -0,0 +1,87 @@
...
+@description('Required. App service plan sku.')
+param sku string = 'IsolatedV2'
+
+@description('Required. App service plan sku code.')
+param skuCode string = 'I1V2'
+
+@description('Required. App service plan worker size.')
+param workerSize string = '6'
+
+@description('Required. App service plan worker size id.')
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:25]: remove this parameter and it's use below

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep [hunk 37]
```diff
@@ -0,0 +1,87 @@
...
+      netFrameworkVersion: netFrameworkVersion
+      alwaysOn: alwaysOn
+    }
+    serverFarmId: hostingPlan.id
+    clientAffinityEnabled: true
+    hostingEnvironmentProfile: {
+      id: resourceId('Microsoft.Web/hostingEnvironments', hostingEnvironmentProfileName)
+    }
+  }
+  dependsOn: [
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:66]: remove dependsOn since you have a reference to the hostingPlan.id

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep [hunk 38]
```diff
@@ -0,0 +1,87 @@
...
+  properties: {
+    name: hostingPlanName
+    workerSize: workerSize
+    workerSizeId: workerSizeId
+    numberOfWorkers: numberOfWorkers
+    hostingEnvironmentProfile: {
+      id: resourceId('Microsoft.Web/hostingEnvironments', hostingEnvironmentProfileName)
+    }
+  }
+  sku: {
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:83]: move this above properties, change property names to lower case

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/subnet.bicep [hunk 39]
```diff
@@ -0,0 +1,23 @@
+@description('Required. The Virtual Network (vNet) Name.')
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/subnet.bicep:1]: remove defaultValues from the parameters since you prompt for them in main.bicep

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep (hunk 29 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/main.bicep]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/privatednszone.bicep (hunk 30 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/privatednszone.bicep:10]: tried to use existing syntax for this, however, it seems not able to get internalInboundIpAddresses[0] property. Error as below,

The type "Microsoft.Web/hostingEnvironments/configurations" does not contain property "internalInboundIpAddresses". Available properties include "apiVersion", "eTag", "extendedLocation", "id", "identity", "kind", "location", "managedBy", "managedByExtended", "name", "plan", "properties", "scale", "sku", "tags", "type", "zones".bicep(BCP053)

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep (hunk 33 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:8]: change value to 'GEN-UNIQUE'

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep (hunk 34 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:11]: removed.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep (hunk 35 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:13]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep (hunk 31 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:48]: removed variables.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep (hunk 37 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:66]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep (hunk 38 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:83]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/subnet.bicep (hunk 39 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/subnet.bicep:1]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep (hunk 32 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:2]: done.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep (hunk 36 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:25]: removed.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/privatednszone.bicep (hunk 30 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/privatednszone.bicep:10]: @bmoore-msft, please advise how to make this change? When I change to the existing syntax, `aseConfig.internalInboundIpAddresses[0]` is not available, I can only leave the old syntax on this. Thanks!

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep [hunk 40]
```diff
@@ -1,58 +1,35 @@
-@description('Required. App service prefix.')
-param appNamePrefix string = 'app-2'
+@description('App service prefix.')
+param appName string = 'GEN-UNIQUE'
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:2]: GEN values are only valid in the parameters file, not in the template - just remove the defaults since this is a module

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep [hunk 41]
```diff
@@ -1,58 +1,35 @@
...
 
-@description('Required. App service location.')
+@description('App service location.')
 param location string = resourceGroup().location
 
-@description('Required. App service plan prefix.')
-param hostingPlanNamePrefix string = 'asev3-asp-2'
+@description('App service plan prefix.')
+param hostingPlanName string = 'GEN-UNIQUE'
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:8]: GEN values are only valid in the parameters file, not in the template - just remove the defaults since this is a module

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep (hunk 40 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep]: done

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep (hunk 41 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep]: done

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/privatednszone.bicep (hunk 30 above)
Comment (bmoore-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/privatednszone.bicep:10]: try:

```aseConfig.properties.internalInboundIpAddresses[0]```

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep (hunk 38 above)
Comment (bmoore-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep:53]: move the sku property above the properties property 
See: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#sort-order-of-properties for the proper sort order of properties

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/privatednszone.bicep (hunk 30 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/privatednszone.bicep]: It works. Thanks.

File: quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep (hunk 38 above)
Comment (samlin-msft) [reply quickstarts/microsoft.web/web-app-asp-app-on-asev3-create/modules/site.bicep]: done.
